egeria#2595 - Add mechanism to retry cohort initialization in the event of failure

jnhagelberg opened the pull request:
See odpi/egeria#2595

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-implementation/pom.xml:
```diff
@@ -41,6 +41,11 @@
             <artifactId>slf4j-api</artifactId>
         </dependency>
 
+        <dependency>
```
(Note that guava is already being used in other places in egeria -- it was just not being used in this particular project)

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/metadatahighway/OMRSMetadataHighwayManager.java:
```diff
@@ -182,23 +182,26 @@ public CohortConnectionStatus connectToCohort(CohortConfig         cohortConfig)
 
             /*
-             * The cohort manager is only added to the list if it initializes successfully.
+             * The cohort manager is always added to the list.
```
Note: this is not a change in behavior.  In the existing code, exceptions were being caught and logged (and not rethrown) in a catch(Throwable) block.  I'm just fixing the comment to reflect what is (and was) actually happening.  I think having it in the list make sense to that the apis to get the cohort status will show cohorts that are still in the connecting/retrying/etc state.

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-apis/src/main/java/org/odpi/openmetadata/repositoryservices/properties/OMRSConfiguration.java:
```diff
@@ -0,0 +1,95 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+/* Copyright Contributors to the ODPi Egeria project. */
+package org.odpi.openmetadata.repositoryservices.properties;
```
@mandy-chessell I'm not sure if that makes sense, since these properties are not specific to any particular cohort.  To minimize the performance impact, there is just one thread pool for the retries that is shared by all of the cohorts.  We could make the number of initialization attempts and the retry interval cohort-specific, but right now those are also shared by all cohorts.

jnhagelberg replied:
I could change the design to have each cohort have its own background thread that handles any initialization retries needed for that particular cohort.  I just thought it would be more efficient for the cohorts to share one thread pool.  I guess that would not be terrible, since initialization retries are not common and there are likely not a large number of cohorts.  I'll look at that.

jnhagelberg replied:
@mandy-chessell So, should there be an initialization retry thread pool that is shared by the cohorts or not .  I am happy to implement it either way, but if there is one, the number of threads in that thread pools needs to configured at the server level.  That is what this class was created for.  If we don't have a thread pool, then the threads will need to be created on demand.  If there are a large number of cohorts trying to connect to kafka, there could potentially be a large number of background threads doing retries.  It seems better to have a a thread pool of a fixed size that handles all of the retries rather than an unbounded number of background threads .  I understand, though, that this thread pool is at the server level, not at the cohort level.  Please advise.

mandy-chessell replied:
It probably needs to be on a cohort by cohort basis since one cohort could successfully initialize and another could fail.

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-implementation/pom.xml:
```diff
@@ -41,6 +41,11 @@
 
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
```
The new code uses the `ImmutableList` class from Guava.  It is not strictly necessary, I can remove it.  I assumed that since it was already being used it had been approved.

planetf1 replied:
Guava gets used transitively by some test tools including testng. It's also used by janusgraph and cassandra. We have to explicitly specify the version to resolve version conflicts. Currently it is also used explicitly in open lineage services

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/metadatahighway/OMRSCohortManager.java:
```diff
@@ -275,12 +302,71 @@ public void initialize(String                           cohortName,
+     * 
+     * Package private so it can be called by {@link OMRSCohortManagerInitializer}
+     */
+    synchronized void setStatus(CohortConnectionStatus status) {
```
? Please clarify.  Is there a formatting document I should be looking at?

jnhagelberg replied:
Oh - I guess for this one it is the opening brace placement.  I'll fix it

jnhagelberg commented on open-metadata-implementation/repository-services/repository-services-implementation/src/main/java/org/odpi/openmetadata/repositoryservices/metadatahighway/OMRSCohortManager.java:
```diff
@@ -275,12 +302,71 @@ public void initialize(String                           cohortName,
+        }
+    }
+
+    private static void unregisterCohortEventListener(
```
@mandy-chessell This method was static as an optimization to avoid to caller needing to pass an implicit *this*  reference that would be unused.  This allows the compiler to do additional optimizations.  However, it most likely little difference in practice, so I'm fine with removing it in this case.  In general I try to make methods static whenever they can be since very little bit helps.

jnhagelberg commented:
@mandy-chessell I have addressed all of your review comments locally.  I have moved the configuration of the max number of retries and the retry interval into the cohort configuration.  Right now I have 2 versions of `OMRSCohortManagerInitializer` -- one which uses a thread pool and one that does not.  

The version that uses a thread pool uses a stripped down version of the `OMRSConfiguration` class here -- now that class only has one property, which is the size of the thread pool.  The behavior in this implementation is that if there is an initialization failure, a task is added to the thread pool that will retry the initialization after the retry interval has elapsed (10 seconds by default).  Also, by default this thread pool will only contain one thread.  The size of the thread pool is independent of the number of cohorts.  They all use this same thread pool for handling their retries.  Basically, what is happening is that there is a queue of initialization retry requests (which is managed by the `ScheduledExecutorService`), and the retry requests are assigned to some thread once a thread is available and its retry interval has elapsed.  My view is that that this does nothing to lessen the independence of the cohort managers.  To me this seems no different than letting different threads handle incoming requests in a web server.  It's the same idea.  The only difference is that in this case it is a queue of retry requests instead of a queue of http requests.

The other version does not use `OMRSConfiguration` configuration at all.  Instead of there being a thread pool, when there is an initialization failure, a new thread is created and started that is specific to the cohort being initialized.  The thread will run until either either the initialization is successful or maximum number of initialization attempts has been exceeded.  At that point, the thread exits.  In the worst case, there could potentially be one retry thread running per cohort.

Anyways, I think I've made my thoughts clear.  If you still think we should remove thread pool mechanism, we'll do that and move on :-)

jnhagelberg commented:
@mandy-chessell @planetf1 ^^

planetf1 commented:
A general comment.... in the scenarios ( #2119 )where I saw a split cohort -- ie multiple members thinking they were in a cohort, but not seeing each other -- there was no initialization failure as such. At least not in terms of api calls etc.

Rather it seemed to be down to the timing of when listeners and consumers started/read kafka events. There's some discussion of this in #2547 (conclusion is to document)

So I'm not clear this change will help with that.
@jnhagelberg did you see a different scenario causing cohort init issues?

I'm also aware we have 2 other related issues
 a) A suggestion of a cohort heartbeat in #2475 @grahamwallis 
 b) Improvements to kafka retries in #1876 

My gut feel is we should focus on b) above first as a more generic level of resiliance for pub/sub, and consider the heartbeat. the fix here may add complexity, and in any case may be best deferred.

@wbittles what's the state of #1876 now?

jnhagelberg commented:
@planetf1 Thanks for the review.  Unfortunately this code here is stale at this point, since I'm still waiting to hear back from Mandy with regard to the thread pool.  Since it has basically been a week since I started getting feedback on that, at this point I think I will just update the PR with the version that uses the thread pool.

jnhagelberg commented:
@planetf1 @mandy-chessell I have addressed your review comments.  

This PR was created to fix a specific issue, which was that if kafka is not available when OMRS starts there was no ability to recover from the error if kafka comes up later.  That is the only issue that I saw.  This usually occurred in conjunction with other errors (for example in the kafka pod).  We would fix the problem in the kafka pod, and then we would have to restart our pod since OMRS had failed to start.  With these changes, now if kafka comes up the system has the ability to recover automatically.

The fix here is specifically related to the initialization of the cohort.  It should have little or no impact on kafka retries.  The PR does not actually change any files in the `kafka-open-metadata-topic-connector` project, which has all of the logic that directly interacts with Kafka.  Any changes to add retries of specific kafka operations would need to go in classes in there, most likely in `KafkaOpenMetadataEventConsumer` or `KafkaOpenMetadataEventProducer`

I took a look at the other issues.  While they are all related to Kafka, they are different issues and should be fixed in their own PRs.

jnhagelberg commented:
@planetf1 One other general comment -- not necessarily about this specific PR, but as I am sure you are aware, the team I am on has limited resources.  We only create egeria PRs for problems we have seen that we have no other way to fix.  We do so not only to make egeria better, but also because the issues create problems in our product that are serious enough to warrant investing the time to not just create an issue for, but to actually investigate and create a PR to fix.  In this case, without the fix egeria does not start at all in some scenarios.  For us, that is a blocking problem, since our code relies very heavily on egeria.  Hopefully you can understand my concern when you come to the conclusion that fixes we have implemented need to be deferred.

I understand that this PR adds some additional complexity.  We have made an attempt to make the fix a simple as possible, but nothing involving doing retries in the background is going to be very simple.  We've made an attempt to keep this as simple and performant as possible.  If you have suggestions for a simpler implemention, I would love to hear them.

jnhagelberg commented:
@mandy-chessell I have updated th PR based on the discussion we had on Friday

planetf1 commented:
Note that the build is failing due to javadoc errors

planetf1 commented:
I tried this fix in master (with #2717) by skipping javadoc & running the coco pharma environment locally.
Started with kafka/zookeeper not running
Ran through start, configure notebooks
Then started kafka,zookeeper.

Both with & without this fix, the end result is that each server was unaware of remote cohort members. I'm not sure if #2717 has changed the behaviour - so refer there first.

I would have expected this scenario to work.

I may have breached the macCohort attempts however. This brings up another comment - we have no AUDIT events being reported in the above code, and it may be desirable to log a change of initializing state so it can be understood  (without in future querying)

Debug entries were reported ie
```
  egeria git:(jnhagelberg-jhagelberg-2595-master) ✗ cat /tmp/log.withjeff | grep 'Initialization of cohort'                                                                <<<
2020-03-06 15:49:29.659  INFO 29968 --- [nio-8080-exec-2] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:29.718  INFO 29968 --- [nio-8080-exec-2] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort devCohort succeeded.
2020-03-06 15:49:29.776  INFO 29968 --- [nio-8080-exec-2] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort iotCohort succeeded.
2020-03-06 15:49:33.046  INFO 29968 --- [nio-8080-exec-8] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:34.871  INFO 29968 --- [nio-8080-exec-4] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:35.960  INFO 29968 --- [io-8080-exec-10] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:36.085  INFO 29968 --- [io-8080-exec-10] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort iotCohort succeeded.
2020-03-06 15:49:38.800  INFO 29969 --- [nio-8081-exec-7] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:41.497  INFO 29969 --- [nio-8081-exec-2] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort cocoCohort succeeded.
2020-03-06 15:49:43.066  INFO 29970 --- [nio-8082-exec-9] o.o.o.r.m.OMRSCohortManagerInitializer   : Initialization of cohort devCohort succeeded.
```
but this is not sufficient in a operational environment.

Full log at https://gist.github.com/ef9f8a7485f05dd054610b1272864453

planetf1 commented:
I backed off #2717 (reverts) and had the same result - ie
 - each cohort member thinks it's in the correct cohorts
 - no cohort member is aware of any other cohort member

jnhagelberg commented:
@planetf1 I just pushed changes to fix the javadoc issue

planetf1 commented:
Note that the cohort initialization succeeds in all cases - even when kafka is down, which is why in my test this logic didn't result in any change. I'm presuming you have a different form of failure . (you said DNS?) which may result in a different path?

jnhagelberg commented:
@planetf1 Yes, in my case the initial attempt to connect to kafka fails.  This PR addresses when that happens and ensures that we retry the connection to kafka is successful.

jnhagelberg commented:
@planetf1 I tested the changes by bringing down my local network connection at the point where the code tries to connect to kafka, while stepping through the code in a debugger.  I verified that after the initial failure, the retry logic kicked in and it was able to successfully connect to kafka.

jnhagelberg commented:
Re 
```
each cohort member thinks it's in the correct cohorts
no cohort member is aware of any other cohort member
```
That sounds like some problem with the initialization protocol.  This PR will do nothing to address that.  This PR only ensures that Egeria eventually connects to Kafka if it is initially down.  In your scenario, if I understand it correctly, kafka *is* available, but the cohort members do not know about each other because of some breakdown in the protocol that gives them that information.

planetf1 commented:
I wasn't 100% sure on the dynamics. However in my case kafka was down and Egeria was able to initilaize in the same way without this fix.

It may just be the exact failure you had originally, to cause cohort registration to fail, is not the situation I have (I had thought it might fail based on what you'd seen - hence the test)

Do you have any more info on that scenario?

Was it that just the hostname of the kafka server was not known?

We would need audit entries IMO.

jnhagelberg commented:
@planetf1 Yes, the only issue in the case that I saw was that the kafka hostname could not initially be resolved.  I will add more audit entries when I get a chance.  Can that be added in a follow on PR?  I am kind of overwhelmed with other things at the moment.

planetf1 commented:
Confirmed that host not found does indeed cause a startup exception for me. Details for reference
```
2020-03-12 09:00:16.377 ERROR 93307 --- [nio-8082-exec-2] o.o.o.r.m.OMRSCohortManager              : Unable to initialize cohort manager

org.apache.kafka.common.KafkaException: Failed to construct kafka consumer
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:820) ~[kafka-clients-2.4.0.jar!/:na]
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:666) ~[kafka-clients-2.4.0.jar!/:na]
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:646) ~[kafka-clients-2.4.0.jar!/:na]
	at org.odpi.openmetadata.adapters.eventbus.topic.kafka.KafkaOpenMetadataEventConsumer.<init>(KafkaOpenMetadataEventConsumer.java:88) ~[kafka-open-metadata-topic-connector-1.6-SNAPSHOT.jar!/:na]
	at org.odpi.openmetadata.adapters.eventbus.topic.kafka.KafkaOpenMetadataTopicConnector.start(KafkaOpenMetadataTopicConnector.java:243) ~[kafka-open-metadata-topic-connector-1.6-SNAPSHOT.jar!/:na]
	at org.odpi.openmetadata.repositoryservices.connectors.omrstopic.OMRSTopicConnector.start(OMRSTopicConnector.java:281) ~[repository-services-apis-1.6-SNAPSHOT.jar!/:na]
```
...
```
Caused by: org.apache.kafka.common.config.ConfigException: Invalid url in bootstrap.servers: newnode
	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:58) ~[kafka-clients-2.4.0.jar!/:na]
	at org.apache.kafka.clients.ClientUtils.parseAndValidateAddresses(ClientUtils.java:47) ~[kafka-clients-2.4.0.jar!/:na]
	at org.apache.kafka.clients.consumer.KafkaConsumer.<init>(KafkaConsumer.java:735) ~[kafka-clients-2.4.0.jar!/:na]
	... 63 common frames omitted
```
...
```
Thu Mar 12 09:00:16 GMT 2020 cocoMDSx Exception OMRS-AUDIT-0006 Configuration error detected while connecting to cohort devCohort, exception org.apache.kafka.common.KafkaException was caught with error message: Failed to construct kafka consumer
```
Full log posted to https://gist.github.com/68078c4a1a920e1d0864441bf733abf9

planetf1 commented:
The above was an example of a kafka send failing - though note it was an OMAS.

I then applied the fix, and noted that the cohort initialation is retried -- for a limited period. Log messages are created, but not audit entries.

Also servers still fail to start if OMASs are configured - as there is no retry there yet. Specifically ASSET_CONSUMER fails to start in the coco environment -- probably because it is first in the list alphabetically...

I wasn't able to get a retry to subsequently work, but that's probably a consequence of my environment

So omrs repos may start but anything more complex will likely fail.

As such the fix here will need extending
 - with audit messages
 - also considering how we handle OMASs
 - more configuration of retries

The main question is whether we are ok to merge this as an incremental improvement, and create a followup (or multiple) to correct these issues., I would suggest we do merge. 
What do you think @mandy-chessell @wbittles ?

wbittles commented:
I'm not sure what the time pressures are like for @planetf1 , but if it would help, the following suggestions might help ? 

kafka provides an admin interface 
[https://kafka.apache.org/23/javadoc/index.html?org/apache/kafka/clients/admin/AdminClient.html](url)
which allows us to list what nodes are available. this would allow the KafkaTopicConnector to poll until the node becomes available 
OR /AND
I could add an afterExecute handler for the consumer which would recreate the consumer, just like the producer has now.

planetf1 commented:
@wbittles - I would regard the details of improving the connector itself as a separate issue (I'd be inclined to consider the latter, but skip the former) - probably best discussed in a connector specific issue.

My point about the OMASs was an observation that we *could* delegate to a new issue and go with a partial fix here... it depends on what we define as success from an initialization perspective. But servers in the cohort will not come up ok if they have OMASs configured without further changes.

The second point is that whilst we have log entries here, we need AUDIT messages since logging is set to OFF by default in Egeria - intentionally. 

The final point is that currently we only try for a short while - but the problem could persist for minutes, or more.. so if we're looking for a general purpose improvement to resiliance, I would default to forever (with audit log reminders - and not a tight loop!) with the option of setting lower.

jnhagelberg commented:
@mandy-chessell What is needed to get this merged?  I've rebased the changes a few times.

mandy-chessell commented:
I am tempted to let this go in as a temporary fix and try to get a more permanent fix in place.  The issue is that it changes the responsiblity for retrying to send the messages from the event management module to the cohort manager.  The reason the original design put the retries with the event manager was that it was able to buffer both the cohort registration events and the type/instance events from the local repository.   The event manager retry stopped working when the futures fix went in because the exceptions changed.

If we move the retry to the cohort manager, we should remove the retry logic from the event manager rather than fix it.  This means we will no longer buffer type and instance events if kafka is down.  Since this is no longer working I am assuming you are ok with this?  It also has the advantage that the server will not bloat with unsent events if the outage is a prolonged one.

Other changes that are needed once this has gone in are to remove the limited retry and extend the REST APIs that can query the cohort status to be able to connect and disconnect the cohort.

I would go with the single thread per cohort for the retry since this is the simplest design.

planetf1 commented:
I would support the idea of removing the retry in the event manager. Queueing can be implemented by the actual topic connector & deployment configuration. For example if using kafka buffer.memory controls the max memory used for the producer queueing - and there are some further timeouts (max.block.ms & request.timeout.ms). 

For the cohort manager I suggest a followon pr that will address a) audit log messages and b) allowing a longer retry period. More broadly after this we need to work through ensuring all the necessary situations are captured - for example the fix here works for host not found, but not necessarily for all cases

jnhagelberg commented:
@mandy-chessell I'm not understanding why you see this as a temporary fix.  I also don't see this as changing responsibility.  The `OMRSMetadataHighwayManager` is still ultimately responsible for initializing the `OMRSCohortManager`.  The The basic change is that now, if the initialization of `OMRSCohortManager` fails in certain ways, the initialization is rerun in the background until it either succeeds or run out of retries.

I guess an alternative to the the approach I took would be to move the retry logic to within  `KafkaOpenMetadataTopicConnector.start()`.  That would be a lot simpler, since this fix is really targeting the use case where the connection to Kafka fails.  Would you be more comfortable with that approach?   With this approach, there would be no additional initialization states or retry threads, just retry looks around the creation of the producer and consumer threads.

I acknowledge that this is not a perfect fix, and there are probably initialization failure scenarios that it does not cover.  It does cover a scenario we have seen, though.

@planetf1 The retry period is already configurable.  I plan to create a follow on PR to add additional audit log messages.

planetf1 commented:
> @planetf1 The retry period is already configurable. I plan to create a follow on PR to add additional audit log messages.
Super - I see that in the code now. must have missed before .

jnhagelberg commented:
@mandy-chessell @planetf1 Ok, I *think* I have addressed your comments.  I have pushed the retries down a level, into the kafka connector.  If we are unable to create either the kafka consumer or producer, the retry logic kicks in.  It still work the same way, with the retries happening in the background.  That way, other things will not be blocked while we are doing the retries.

I have also added audit logging.  I turned the initialization retry mechanism into a reusable component that can be used to do retries at any level or for any component if we so choose.

jnhagelberg commented:
Note - I still need to do some additional testing to make sure that this has not introduced regressions in the handling of the problem use case we're seeing.  It's pretty late here, I'll try that tomorrow.

jnhagelberg commented:
Ok, @mandy-chessell @planetf1 today I was able to verify that the fix still works for the use case where the kafka host is initially unavailable.

jnhagelberg commented:
@mandy-chessell @planetf1 Any questions/comments?

planetf1 commented:
@wbittles Can you elaborate on the plan for the retries?

wbittles commented:
@jnhagelberg @planetf1  @mandy-chessell I've been looking at other Kafka connectivity issues, some transient and recoverable and others that will require an intervention. My main concern is around error handling. To solve my own issues I've suggested a fix that should as a consequence make this change unnecessary.

The fix I've suggested requires the connector to wait until Kafka achieves full "bring up", before Egeria begins making connections. So if the KafkaTopicConnector waits until Kafka reports that the Topic is available before we establish the producers and consumers. It still needs testing but the idea is that by waiting for a high level event like making the Topic available we can avoid having to implement handler for each of the individual bring up issues that is reported. 

Waiting for high level events will also simplify the de-duplication of error reporting when the error reported is transient. I hope to open a PR for the new change this week for testing, but in theory it should help simplify the Egeria code base.

jnhagelberg commented:
@wbittles Can you please provide a pointer to that fix, when it is ready?  I'd like to verify that this issue is resolved with your fix.

wbittles commented:
@jnhagelberg #2990 is the PR for the change, and the verification would be really appreciated.

planetf1 commented:
Set as draft pending check of #2990 - it's a different fix but I believe it does address an issue with host not found during startup of egeria  - not just in cohort registration, but more generally.

planetf1 commented:
I'm going to close this PR for now as I hope the latest fix will address your notes. We can reopen the issue/PR if needed.
